Add vector search documentation

kolchfa-aws opened the pull request:
Adds a vector search section

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

naveentatikonda commented on _vector-search/vector-search-techniques/index.md:
```diff
@@ -0,0 +1,68 @@
+|:---|:---|:---|:---|:---|
+|  Max dimensions |  16,000  |  16,000 |  16,000 |  16,000 |
+|  Filter |  Post-filter |  Post-filter |  Post-filter |  Filter during search |
+|  Training required |  No |  No |  Yes |  No |
```
Faiss HNSW with PQ also requires training

naveentatikonda commented on _vector-search/optimizing-performance/knn-vector-quantization.md:
```diff
@@ -0,0 +1,41 @@
+redirect_from:
+  - /search-plugins/knn/knn-vector-quantization/
+outside_cards:
+  - heading: "Byte vectors"
```
Need to add a card for binary vectors along with byte vectors
https://opensearch.org/docs/latest/field-types/supported-field-types/knn-vector#binary-vectors

kolchfa-aws replied:
@naveentatikonda Thanks! I addressed both comments. Could you review this commit https://github.com/opensearch-project/documentation-website/pull/9135/commits/a5e8b8db410d5d6538bc88abd5df5d82827d3494.

naveentatikonda replied:
@kolchfa-aws changes looks good. Thanks for making those changes. 

For binary vectors, we also need to add memory estimation. Here is the formula for HNSW
`1.1 * (dimension / 8 + 8 * M) bytes/vector`

For IVF, I guess it is `1.1 * (((dimension / 8) * num_vectors) + (nlist * dimension / 8))`. @jmazanec15 can you pls confirm ?

jmazanec15 replied:
Yeah that looks good to me

kolchfa-aws replied:
Added both formulas.

jmazanec15 commented:
Hi @kolchfa-aws, this looks awesome!

In general, I think we should start moving the more low-level/expert details (like quantization and method configuration), out of vector search section and into detailed field reference section. Here is some high level feedback:
1. Itd be good to showcase/highlight some high level features on the vector search splash page (https://kolchfa-aws.github.io/vector-search). In particular: filtering, multi-vector per document support (nested), automatic embedding generation, low-memory search, and hybrid search
2. For vector search techniques, can we add sections on sparse vectors and hybrid search?
3. Can we add knn query type in query dsl?
4. Can we have a page for space types and only maintain one table - possibly in field reference? Also, can we move engine/method details into field reference and each engine have its own detailed section in mapping reference.
5. For examples/tutorials, can we use basic mapping instead of complex mapping:
```
PUT /test-index
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "my_vector1": {
        "type": "knn_vector",
        "dimension": 3,
        "space_type": "l2"
      }
    }
  }
}
```
In performance tuning, we can mention picking a specific engine or specifying overriding method parameters for expert level fine tuning and point to reference docs.
6. For quantization, can we put low level details into an expert section? Instead, can we just mention in performance tuning for memory optimization, we use quantization to achieve 32x, 16x, 8x, 4x, 2x compression_levels. I think we can even renaming the disk-based section, memory optimized vector search. This is example mapping:
```
PUT my-vector-index
{
  "settings" : {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "my_vector_field": {
        "type": "knn_vector",
        "dimension": 8,
        "space_type": "innerproduct",
        "data_type": "float",
        "mode": "on_disk",
        "compression_level": "16x"
      }
    }
  }
}
```
Quantization is a bit ugly for users to have to understand. So I think its better to belong in detailed field reference. We can say, for further fine tuning of the quantization methods, see field reference.
7. Can we move SIMD optimization section to performance tuning: https://kolchfa-aws.github.io/vector-search/creating-vector-index/vector-field/#simd-optimization-for-the-faiss-engine? End users dont really need to know about this when reading about vector data types
8. In optimizing vector search performance, can we add page on cluster sizing? Basically, this should point to memory evaluation formulas and picking node configurations.

kolchfa-aws commented:
@jmazanec15 Thank you for the review; I addressed your feedback. Please take a look again when you get a chance.

jmazanec15 commented on _vector-search/index.md:
```diff
@@ -0,0 +1,30 @@
+
+Vector search enhances traditional lexical search by encoding text, images, and audio as vectors, enabling similarity search with semantic understanding. OpenSearch unifies traditional search, analytics, and vector search in a single solution. As a vector database, it streamlines AI application development by efficiently storing and retrieving high-dimensional data.
+
+<span class="centering-container">
```
Can we move button below key features? It looks a bit odd in the middle of the page/

jmazanec15 commented on _vector-search/getting-started/vector-search-options.md:
```diff
@@ -0,0 +1,94 @@
+    link: "/vector-search/searching-data/#searching-auto-generated-embeddings"
+---
+
+# Bringing or generating embeddings
```
Can this page be moved into Vector Search Basics? Im not sure it needs its own page.

kolchfa-aws replied:
Having this as a separate page allows users to discover this page while scanning the left nav. Vector search basics provides info about vector search for beginners: experienced users might not look there. This page provides info about the options they have, so I think it's warranted as a separate page.

jmazanec15 commented on _vector-search/vector-search-techniques/index.md:
```diff
@@ -0,0 +1,72 @@
+
+For more information, see [Methods and engines]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/) and [Choosing the right method]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#choosing-the-right-method).
+
+## Engine recommendations
```
Lets remove this section - for more details, users can reference the methods and engines section.

kolchfa-aws replied:
@jmazanec15 Should we move this into the "Methods and engines" page? This table has information about things not mentioned in the "Methods and engines", like filter options and query latency and quality.

jmazanec15 replied:
Right, I think it makes sense to be moved to that page

jmazanec15 commented on _vector-search/vector-search-techniques/index.md:
```diff
@@ -0,0 +1,72 @@
+
+## Approximate search
+
+OpenSearch supports the following algorithms (_methods_) for approximate vector search:
```
Can we shorten this section to just point to the engines/methods docs? We should just say that OpenSearch supports a variety of different backend engines and algorithms. Based on your mode and memory requirements, OpenSearch will select the best default configuration.

jmazanec15 commented on _vector-search/vector-search-techniques/index.md:
```diff
@@ -0,0 +1,72 @@
+
+- [Approximate search]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/)) (approximate k-NN, or ANN): Returns approximate nearest neighbors to the query vector. Usually, approximate search algorithms sacrifice indexing speed and search accuracy in exchange for performance benefits such as lower latency, smaller memory footprints, and more scalable search. For most use cases, appro ...
+
+- Exact search: A brute-force, exact k-NN search of vector fields. OpenSearch supports the following types of exact search: 
```
Should we add a brief section on exact search for symmetry with ANN?

kolchfa-aws replied:
@jmazanec15 Could you please provide the information to create this section?

kolchfa-aws commented on _query-dsl/specialized/kNN.md:
```diff
@@ -0,0 +1,208 @@
+
+Field | Data type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`vector` | Array of floats | Required | The query vector to use for vector search.
```
Reworded.

kolchfa-aws commented:
@jmazanec15 I addressed your comments in this commit https://github.com/opensearch-project/documentation-website/pull/9135/commits/79eeff141ff7700999083a880fead20aa935e7be. Could you take a look again?

natebower commented on _vector-search/vector-search-techniques/painless-functions.md:
```diff
@@ -0,0 +1,80 @@
+layout: default
+title: Painless extensions
+nav_order: 25
+parent: Exact k-NN search with scoring script
```
with "a" scoring script?

natebower commented on _vector-search/vector-search-techniques/knn-score-script.md:
```diff
@@ -102,9 +105,9 @@ All parameters are required.
 
-The [post filter example in the approximate approach]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn#using-approximate-k-nn-with-filters) shows a search that returns fewer than `k` results. If you want to avoid this situation, the score script method lets you essentially invert the order of events. In other words, you can filter down the set of documents over which to execute the ...
+The [post filter example in the approximate approach]({{site.url}}{{site.baseurl}}/vector-search/filter-search-knn/) shows a search that returns fewer than `k` results. If you want to avoid this situation, the score script method lets you essentially invert the order of events. In other words, you can filter down the set of documents over which to execute the k-nearest neighbor search.
 
```
Global: Ensure we define k-NN on first appearance and use k-NN thereafter. On this line, confirm the accuracy of the link text.

natebower commented on _vector-search/vector-search-techniques/painless-functions.md:
```diff
@@ -0,0 +1,80 @@
+
+With the Painless scripting extensions, you can use k-NN distance functions directly in your Painless scripts to perform operations on `knn_vector` fields. Painless has a strict list of allowed functions and classes per context to ensure its scripts are secure. OpenSearch adds Painless Scripting extensions to a few of the distance functions used in [k-NN score script]({{site.url}}{{site.baseurl}} ...
+
+## Get started with k-NN's Painless Scripting functions
```
```suggestion
## Get started with k-NN Painless scripting functions
```

natebower commented on _vector-search/vector-search-techniques/painless-functions.md:
````diff
@@ -0,0 +1,80 @@
+   "source": "doc[params.field].size() == 0 ? 0 : 1 / (1 + l2Squared(params.query_value, doc[params.field]))",
+   ```
+
+   Because scores can only be positive, this script ranks documents with vector fields higher than those without.
````
```suggestion
   Because scores can only be positive, this script ranks documents with vector fields higher than those without vector fields.
```

natebower commented on _vector-search/vector-search-techniques/painless-functions.md:
```diff
@@ -0,0 +1,80 @@
+The following table describes the Painless functions OpenSearch provides.
+
+Function name | Function signature | Description
+:--- | :---
```
Line 57, last sentence: `divide by 0` errors?

natebower commented on _vector-search/vector-search-techniques/painless-functions.md:
```diff
@@ -0,0 +1,80 @@
+
+   Because scores can only be positive, this script ranks documents with vector fields higher than those without.
+
+With cosine similarity, it is not valid to pass a zero vector (`[0, 0, ...]`) as input. This is because the magnitude of such a vector is 0, which raises a `divide by 0` exception in the corresponding formula. Requests containing the zero vector will be rejected, and a corresponding exception will be thrown.
```
"With cosine similarity" => "When using cosine similarity"?

kolchfa-aws replied:
```suggestion
When using cosine similarity, it is not valid to pass a zero vector (`[0, 0, ...]`) as input. This is because the magnitude of such a vector is 0, which raises a `divide by 0` exception in the corresponding formula. Requests containing the zero vector will be rejected, and a corresponding exception will be thrown.
```

natebower reviewed (commented):
@kolchfa-aws Great job putting all of this together 😄. Please see my comments and changes and let me know if you have any questions. Thanks!

kolchfa-aws commented on _field-types/supported-field-types/knn-memory-optimized.md:
```diff
@@ -0,0 +1,926 @@
+When using `byte` vectors, expect some loss of precision in the recall compared to using `float` vectors. Byte vectors are useful in large-scale applications and use cases that prioritize a reduced memory footprint in exchange for a minimal loss of recall.
+{: .important}
+
+When using `byte` vectors with the `faiss` engine, we recommend using [SIMD optimization]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#simd-optimization), which helps to significantly reduce search latencies and improve indexing throughput.
```
```suggestion
When using `byte` vectors with the `faiss` engine, we recommend using [Single Instruction Multiple Data (SIMD) optimization]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#simd-optimization), which helps to significantly reduce search latencies and improve indexing throughput.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-memory-optimized.md:
```diff
@@ -0,0 +1,926 @@
+
+### Quantization techniques
+
+If your vectors are of the type `float`, you need to first convert them to the `byte` type before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depen ...
```
```suggestion
If your vectors are of the type `float`, you need to first convert them to the `byte` type before ingesting documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. The Faiss engine supports several quantization techniques, such as scalar quantization (SQ) and product quantization (PQ). The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/opensearch-benchmark-workloads/tree/main/vectorsearch) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types. The provided pseudocode is for illustration purposes only.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+- The `engine` that will implement the method (for example, `faiss` or `lucene`)
+- A map of `parameters` specific to that implementation
+
+The following example configures an `hnsw` method with an `l2` space type, a `faiss` engine, and the method-specific parameters:
```
```suggestion
The following example configures an `hnsw` method with the `l2` space type, the `faiss` engine, and the method-specific parameters:
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
````diff
@@ -0,0 +1,422 @@
+```
+{% include copy-curl.html %}
+
+Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the specific engine section.
````
```suggestion
Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the section for a specific engine.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+
+Mapping parameter | Required | Default | Updatable | Description
+:--- | :--- | :--- | :--- | :---
+`name` | Yes | N/A | No | The nearest neighbor method. Valid values are `hnsw` and `ivf`. Not every engine combination supports each of the methods. For a list of supported methods, see the specific engine section.
```
```suggestion
`name` | Yes | N/A | No | The nearest neighbor method. Valid values are `hnsw` and `ivf`. Not every engine combination supports each of the methods. For a list of supported methods, see the section for a specific engine.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+Mapping parameter | Required | Default | Updatable | Description
+:--- | :--- | :--- | :--- | :---
+`name` | Yes | N/A | No | The nearest neighbor method. Valid values are `hnsw` and `ivf`. Not every engine combination supports each of the methods. For a list of supported methods, see the specific engine section.
+`space_type` | No | `l2` | No | The vector space used to calculate the distance between vectors. Valid values are `l1`, `l2`, `linf`, `cosinesimil`, `innerproduct`, `hamming`, and `hammingbit`. Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the specific engine section. Note: This value can also be specified at the top level of the mapping. For ...
```
```suggestion
`space_type` | No | `l2` | No | The vector space used to calculate the distance between vectors. Valid values are `l1`, `l2`, `linf`, `cosinesimil`, `innerproduct`, `hamming`, and `hammingbit`. Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the section for a specific engine. Note: This value can also be specified at the top level of the mapping. For more information, see [Spaces]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-spaces/).
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+`name` | Yes | N/A | No | The nearest neighbor method. Valid values are `hnsw` and `ivf`. Not every engine combination supports each of the methods. For a list of supported methods, see the specific engine section.
+`space_type` | No | `l2` | No | The vector space used to calculate the distance between vectors. Valid values are `l1`, `l2`, `linf`, `cosinesimil`, `innerproduct`, `hamming`, and `hammingbit`. Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the specific engine section. Note: This value can also be specified at the top level of the mapping. For ...
+`engine` | No | `faiss`  | No | The approximate k-NN library to use for indexing and search. Valid values are `faiss`, `lucene`, and `nmslib` (deprecated).
+`parameters` | No | `null` | No | The parameters used for the nearest neighbor method. For more information, see the specific engine section.
```
```suggestion
`parameters` | No | `null` | No | The parameters used for the nearest neighbor method. For more information, see the section for a specific engine.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+:--- | :--- | :--- | :--- | :---
+`nlist` | No | 4 | No | Number of buckets to partition vectors into. Higher values may increase accuracy but increase memory and training latency.
+`nprobes` | No | 1 | No | Number of buckets to search during query. Higher values increase accuracy but slow searches.
+`encoder` | No | flat | No | Encoder definition for encoding vectors.
```
```suggestion
`encoder` | No | flat | No | An encoder definition for encoding vectors.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+Parameter name | Required | Default | Updatable | Description
+:--- | :--- | :--- | :--- | :---
+`ef_construction` | No | 100 | No | The size of the dynamic list used during k-NN graph creation. Higher values result in a more accurate graph but slower indexing speed.
+`m` | No | 16 | No | The number of bidirectional links created for each new element. Impacts memory consumption significantly. Keep between 2 and 100.
```
```suggestion
`m` | No | 16 | No | The number of bidirectional links created for each new element. Impacts memory consumption significantly. Keep between `2` and `100`.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-methods-engines.md:
```diff
@@ -0,0 +1,422 @@
+
+## Choosing the right method
+
+There are several options to choose from when building your `knn_vector` field. To determine the correct methods and parameters, you should first understand the requirements of your workload and what trade-offs you are willing to make. Factors to consider are (1) query latency, (2) query quality, (3) memory limits, and (4) indexing latency.
```
```suggestion
There are several options to choose from when building your `knn_vector` field. To select the correct method and parameters, you should first understand the requirements of your workload and what trade-offs you are willing to make. Factors to consider are (1) query latency, (2) query quality, (3) memory limits, and (4) indexing latency.
```

kolchfa-aws commented on _field-types/supported-field-types/knn-spaces.md:
```diff
@@ -0,0 +1,98 @@
+
+A space defines the function used to measure the distance between two points in order to determine the k-nearest neighbors. From the k-NN perspective, a lower score equates to a closer and better result. This is the opposite of how OpenSearch scores results, where a higher score equates to a better result. OpenSearch supports the following spaces. 
+
+Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the specific engine section in the [method documentation]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/).
```
```suggestion
Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the section for a specific engine in the [method documentation]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/).
```

kolchfa-aws commented on _field-types/supported-field-types/knn-vector.md:
```diff
@@ -222,687 +101,79 @@ PUT test-index
+`type` | String | The vector field type. Must be `knn_vector`. Required.
+`dimension` | Integer | The size of the vectors used. Valid values are in the [1, 16,000] range. Required.
+`data_type` | String | The data type for the vector elements. Valid values are `binary`, `byte`, and `float`. Optional. Default is `float`.
+`space_type` | String | The vector space used to calculate the distance between vectors. Valid values are `l1`, `l2`, `linf`, `cosinesimil`, `innerproduct`, `hamming`, and `hammingbit`. Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the specific engine section. Note: This value can also be specified within the `method`. Optional. For more info ...
```
```suggestion
`space_type` | String | The vector space used to calculate the distance between vectors. Valid values are `l1`, `l2`, `linf`, `cosinesimil`, `innerproduct`, `hamming`, and `hammingbit`. Not every method/engine combination supports each of the spaces. For a list of supported spaces, see the section for a specific engine. Note: This value can also be specified within the `method`. Optional. For more information, see [Spaces]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-spaces/). 
```

kolchfa-aws commented on _includes/home_cards.html:
```diff
@@ -0,0 +1,72 @@
+    <div class="home-card">
+      <a href="{{site.url}}/docs/latest/benchmark/" class='card-link'></a>
+        <p class="heading">OpenSearch Benchmark</p>
+        <p class="description">Measure performance metrics for your OpenSearch cluster.</p>
```
```suggestion
        <p class="description">Measure OpenSearch cluster performance metrics.</p>
```

kolchfa-aws commented on _query-dsl/specialized/k-nn.md:
```diff
@@ -0,0 +1,208 @@
+Field | Data type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`vector` | Array of floats or bytes | Required | The query vector to use for vector search. The data type of the vector elements must match the data type of vectors indexed in the [`knn_vector` field]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/) searched.
+`k` | Integer | Optional | The number of nearest neighbors to return. Valid values are in the [1, 10,000] range. Required if `max_distance` or `min_score` is not specified.
```
```suggestion
`k` | Integer | Optional | The number of nearest neighbors to return. Valid values are in the [1, 10,000] range. Required if either `max_distance` or `min_score` is not specified.
```

kolchfa-aws commented on _query-dsl/specialized/k-nn.md:
```diff
@@ -0,0 +1,208 @@
+`k` | Integer | Optional | The number of nearest neighbors to return. Valid values are in the [1, 10,000] range. Required if `max_distance` or `min_score` is not specified.
+`max_distance` | Float | Optional | The maximum distance threshold for search results. Only one of `k`, `max_distance`, or `min_score` can be specified. For more information, see [Radial search]({{site.url}}{{site.baseurl}}/vector-search/specialized-operations/radial-search-knn/).
+`min_score` | Float | Optional | The minimum score threshold for search results. Only one of `k`, `max_distance`, or `min_score` can be specified. For more information, see [Radial search]({{site.url}}{{site.baseurl}}/vector-search/specialized-operations/radial-search-knn/).
+`filter` | Object | Optional | A filter to apply to the k-NN search. For more information, see [Vector search with filters]({{site.url}}{{site.baseurl}}/vector-search/filter-search-knn/). **Important**: Filter can only be used with the `faiss` or `lucene` engines.
```
```suggestion
`filter` | Object | Optional | A filter to apply to the k-NN search. For more information, see [Vector search with filters]({{site.url}}{{site.baseurl}}/vector-search/filter-search-knn/). **Important**: A filter can only be used with the `faiss` or `lucene` engines.
```

kolchfa-aws commented on _query-dsl/specialized/k-nn.md:
```diff
@@ -0,0 +1,208 @@
+
+## Rescoring results
+
+You can fine-tune search by providing the `ef_search` and `oversample_factor` parameters.
```
```suggestion
You can fine-tune search by providing the `ef_search` and `oversample_factor` parameters.

```

kolchfa-aws commented on _vector-search/creating-vector-index.md:
```diff
@@ -0,0 +1,151 @@
+{% include copy-curl.html %}
+
+
+Regardless of the type of vector search, the following elements are part of creating a vector index:
```
```suggestion
Configuring vector search involves the following key steps:
```

kolchfa-aws commented on _vector-search/creating-vector-index.md:
```diff
@@ -0,0 +1,151 @@
+1. **Define a vector field**:
+   Specify the field that will store the vector data. When defining a `knn_vector` field in OpenSearch, you can select from different data types to balance storage requirements and performance. By default, k-NN vectors are float vectors, but you can also opt for byte or binary vectors for more efficient storage. For more information, see [k-NN vector]({{site.url}}{{site.baseurl}}/field-types/supp ...
+
+1. **Specify dimension**:
```
```suggestion
1. **Specify the dimension**:
```

kolchfa-aws commented on _vector-search/getting-started/vector-search-options.md:
```diff
@@ -0,0 +1,94 @@
+    link: "/vector-search/getting-started/tutorials/neural-search-tutorial/"
+pre_items:
+  - heading: "Generate embeddings"
+    description: "Generate embeddings outside of OpenSearch using your favorite embedding utility."
```
```suggestion
    description: "Generate embeddings outside of OpenSearch using your favorite embedding utility"
```

kolchfa-aws commented on _vector-search/getting-started/vector-search-options.md:
```diff
@@ -0,0 +1,94 @@
+    link: "/vector-search/searching-data/#searching-raw-vectors"
+auto_items:
+  - heading: "Configure an embedding model"
+    description: "Configure a machine learning model that will automatically generate embeddings from your text at ingest time and query time."
```
```suggestion
    description: "Configure a machine learning model that will automatically generate embeddings from your text at ingestion time and query time."
```

kolchfa-aws commented on _vector-search/getting-started/vector-search-options.md:
```diff
@@ -0,0 +1,94 @@
+
+# Bringing your own or generating embeddings
+
+In OpenSearch, you can either bring your own vectors or let OpenSearch generate them automatically from your data. Automated embedding generation integrated into OpenSearch reduces data preprocessing effort at ingestion and search time.
```
```suggestion
In OpenSearch, you can either bring your own vectors or let OpenSearch generate them automatically from your data. Letting OpenSearch automatically generate your embeddings reduces data preprocessing effort at ingestion and search time.
```

kolchfa-aws commented on _vector-search/getting-started/vector-search-options.md:
```diff
@@ -0,0 +1,94 @@
+
+### Option 2: Generate embeddings within OpenSearch
+
+OpenSearch automatically generates vector embeddings from your data using a machine learning (ML) model.
```
```suggestion
Use this option to let OpenSearch automatically generate vector embeddings from your data using a machine learning (ML) model.
```

kolchfa-aws commented on _vector-search/ml-powered-search/index.md:
```diff
@@ -0,0 +1,76 @@
+    link: "/vector-search/ml-powered-search/conversational-search/"
+chunking_cards:
+  - heading: "Text chunking"
+    description: "Use text chunking to ensure adherence to token limit for embedding models."
```
```suggestion
    description: "Use text chunking to ensure adherence to embedding model token limits."
```

kolchfa-aws commented on _vector-search/performance-tuning.md:
```diff
@@ -0,0 +1,37 @@
+This topic provides performance tuning recommendations to improve indexing and search performance for approximate k-NN (ANN). From a high level, k-NN works according to these principles:
+* Vector indexes are created per knn_vector field / (Lucene) segment pair.
+* Queries execute on segments sequentially inside the shard (same as any other OpenSearch query).
+* The coordinator node picks up final size number of neighbors from the neighbors returned by each shard.
```
```suggestion
* The coordinator node selects the final `size` neighbors from the neighbors returned by each shard.
```

kolchfa-aws commented on _vector-search/settings.md:
```diff
@@ -0,0 +1,50 @@
+`knn.cache.item.expiry.enabled` | Dynamic | `false` | Whether to remove native library indexes that have not been accessed for a certain duration from memory.
+`knn.cache.item.expiry.minutes` | Dynamic | `3h` | If enabled, the amount of idle time before a native library index is removed from memory.
+`knn.circuit_breaker.unset.percentage` | Dynamic | `75` | The native memory usage threshold for the circuit breaker. Memory usage must be lower than this percentage of `knn.memory.circuit_breaker.limit` in order for `knn.circuit_breaker.triggered` to remain `false`.
+`knn.circuit_breaker.triggered` | Dynamic | `false` | True when memory usage exceeds the `knn.circuit_breaker.unset.percentage` value.
```
```suggestion
`knn.circuit_breaker.triggered` | Dynamic | `false` | `true` when memory usage exceeds the `knn.circuit_breaker.unset.percentage` value.
```

kolchfa-aws commented on _vector-search/settings.md:
```diff
@@ -0,0 +1,50 @@
+`knn.model.index.number_of_shards`| Dynamic | `1` | The number of shards to use for the model system index, which is the OpenSearch index that stores the models used for approximate nearest neighbor (ANN) search.
+`knn.model.index.number_of_replicas`| Dynamic | `1` | The number of replica shards to use for the model system index. Generally, in a multi-node cluster, this value should be at least 1 in order to increase stability.
+`knn.model.cache.size.limit` | Dynamic | `10%` |  The model cache limit cannot exceed 25% of the JVM heap.
+`knn.faiss.avx2.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx2.so` library and load the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engin ...
```
```suggestion
`knn.faiss.avx2.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx2.so` library and load the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [Single Instruction Multiple Data (SIMD) optimization]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-methods-engines/#simd-optimization).
```

kolchfa-aws commented on _vector-search/settings.md:
```diff
@@ -0,0 +1,50 @@
+
+## Index settings
+
+The following table lists all available index-level k-NN settings. For information about updating these settings, see [Index-level index setting]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/index-settings/#index-level-index-settings).
```
```suggestion
The following table lists all available index-level k-NN settings. For information about updating these settings, see [Index-level index settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/index-settings/#index-level-index-settings).
```

natebower commented on _vector-search/tutorials/neural-search-tutorial.md:
````diff
@@ -0,0 +1,877 @@
+```
+{% include copy-curl.html %}
+
+Document 3 is not returned because it does not contain the specified keywords. Documents containing the words `rodeo` and `cowboy` are scored lower because semantic meaning is not considered:
````
Should "their" precede "semantic meaning"?

natebower commented on _vector-search/tutorials/neural-search-tutorial.md:
```diff
@@ -0,0 +1,877 @@
+
+### Search using a semantic search
+
+To search using semantic search, use a `neural` query and provide the model ID of the model you set up earlier so that vector embeddings for the query text are generated with the model used at ingestion time:
```
Insert "a" before "semantic search".

natebower commented on _vector-search/tutorials/neural-search-tutorial.md:
```diff
@@ -0,0 +1,877 @@
+## Further reading
+
+- Read about the basics of OpenSearch semantic search in [Building a semantic search engine in OpenSearch](https://opensearch.org/blog/semantic-search-solutions/).
+- Read about the benefits of combining keyword and semantic search, the normalization and combination technique options, and benchmarking tests in [The ABCs of semantic search in OpenSearch: Architectures, benchmarks, and combination strategies](https://opensearch.org/blog/semantic-science-benchmarks/).
```
Delete "the benefits of" so that it isn't read as applying to the second and third list items.

natebower commented on _vector-search/tutorials/semantic-search/index.md:
```diff
@@ -5,48 +5,55 @@ parent: Tutorials
+      - "<b>Platform:</b> OpenSearch, Amazon OpenSearch Service"
+      - "<b>Model:</b> Amazon Titan"  
+      - "<b>Deployment:</b> Amazon Bedrock (in a different account than your Amazon OpenSearch Service account)"
+  - heading: Semantic search using a model on Amazon SageMaker
```
"on" => "in"?

natebower commented on _vector-search/vector-search-techniques/index.md:
```diff
@@ -31,8 +31,8 @@ OpenSearch supports multiple backend algorithms (_methods_) and libraries for im
 ## Combining multiple search techniques
 
-_Hybrid search_ enhances search relevance by combining multiple search techniques in OpenSearch. It integrates traditional keyword search with vector-based semantic search. Through a configurable search pipeline, hybrid search normalizes and combines scores from different search methods to provide unified, relevant results. This approach is particularly effective for complex queries where both se ...
+_Hybrid search_ enhances search relevance by combining multiple search techniques in OpenSearch. It integrates traditional keyword search with vector-based semantic search. Through a configurable search pipeline, hybrid search normalizes and combines scores from different search methods to provide unified, relevant results. This approach is particularly effective for complex queries where both se ...
```
I would change "for complex queries where" to "for complex queries in which"

natebower commented on _vector-search/tutorials/index.md:
```diff
@@ -3,130 +3,241 @@ layout: default
+      - "<b>Platform:</b> OpenSearch, Amazon OpenSearch Service"
+      - "<b>Model:</b> Amazon Titan"  
+      - "<b>Deployment:</b> Amazon Bedrock (in a different account than your Amazon OpenSearch Service account)"
+  - heading: Semantic search using a model on Amazon SageMaker
```
"on" => "in"?

natebower reviewed (commented):
@kolchfa-aws Please see my comments and let me know if you have any questions. Thanks!

natebower commented on _vector-search/getting-started/auto-generated-embeddings.md:
```diff
@@ -248,6 +253,70 @@ The response contains the matching documents:
+
+You can quickly set up automatic embedding generation using [_automated workflows_]({{site.url}}{{site.baseurl}}/automating-configurations/). This approach automatically creates and provisions all necessary resources. For more information, see [Workflow templates]({{site.url}}{{site.baseurl}}/automating-configurations/workflow-templates/).
+
+You can use automated workflows to create and deploy externally hosted models and create resources for various AI search types. In this example, you'll create the same search you've created using manual steps.
```
"you've created using" => "you've already created following"

natebower commented on _vector-search/getting-started/auto-generated-embeddings.md:
```diff
@@ -248,6 +253,70 @@ The response contains the matching documents:
+
+### Step 2: Configure a workflow
+
+Create and provision a semantic search workflow. You must provide the model ID for the model you configured. Review your selected workflow template [defaults](https://github.com/opensearch-project/flow-framework/blob/2.13/src/main/resources/defaults/semantic-search-defaults.json) to determine whether you need to update any of the parameters. For example, if the model dimensionality is different f ...
```
"for the model you configured" => "for the configured model"

natebower commented on _vector-search/getting-started/auto-generated-embeddings.md:
````diff
@@ -248,6 +253,70 @@ The response contains the matching documents:
+}
+```
+
+You can now continue with [Steps 4 and 5](#step-4-ingest-documents-into-the-index-ingest-documents-into-the-index) to ingest documents into the index and search the index.
````
When it's plural, I think "steps" needs to be lowercase. Or you could separate them into Step 4 and Step 5.

natebower reviewed (commented):
@kolchfa-aws Three comments on the indicated commit.

natebower reviewed (approved):
LGTM
